common/fitparser.cpp

```diff
@@ -0,0 +1,1090 @@
+        
+        if (found.isValid()) {
+            // Found it, no need to process further
+            return;
+        }
+    }
+    
+    // Check for all FIT signatures in item body
+    UByteArray lastVtfBody = model->body(ffsParser->lastVtf);
+    UINT64 fitSignatureValue = INTEL_FIT_SIGNATURE;
+    UByteArray fitSignature((const char*)&fitSignatureValue, sizeof(fitSignatureValue));
+    UINT32 storedFitAddress = *(const UINT32*)(lastVtfBody.constData() + lastVtfBody.size() - INTEL_FIT_POINTER_OFFSET);
+    for (INT32 offset = (INT32)model->body(index).indexOf(fitSignature);
+         offset >= 0;
+         offset = (INT32)model->body(index).indexOf(fitSignature, offset + 1)) {
+        // FIT candidate found, calculate its physical address
+        UINT32 fitAddress = (UINT32)(model->base(index) + (UINT32)ffsParser->addressDiff + model->header(index).size() + (UINT32)offset);
+        
+        // Check FIT address to be stored in the last VTF
+        if (fitAddress == storedFitAddress) {
+            found = index;
+            fitOffset = offset;
+            msg(usprintf("%s: real FIT table found at physical address %08Xh", __FUNCTION__, fitAddress), found);
+            break;
+        }
+        else if (model->rowCount(index) == 0) { // Show messages only to leaf items
+            msg(usprintf("%s: FIT table candidate found, but not referenced from the last VTF", __FUNCTION__), index);
+        }
+    }
+}
+
+USTATUS FitParser::parseFitEntryMicrocode(const UByteArray & microcode, const UINT32 localOffset, const UModelIndex & parent, UString & info, UINT32 &realSize)
+{
+    U_UNUSED_PARAMETER(parent);
+    if ((UINT32)microcode.size() - localOffset < sizeof(INTEL_MICROCODE_HEADER)) {
+        return U_INVALID_MICROCODE;
+    }
+    
+    const INTEL_MICROCODE_HEADER* ucodeHeader = (const INTEL_MICROCODE_HEADER*)(microcode.constData() + localOffset);
+    if (!ffsParser->microcodeHeaderValid(ucodeHeader)) {
+        return U_INVALID_MICROCODE;
+    }
+    
+    if ((UINT32)microcode.size() - localOffset < ucodeHeader->TotalSize) {
+        return U_INVALID_MICROCODE;
+    }
+    
+    // Valid microcode found
+    info = usprintf("CpuSignature: %08Xh, Revision: %08Xh, Date: %02X.%02X.%04X",
+                    ucodeHeader->ProcessorSignature,
+                    ucodeHeader->UpdateRevision,
+                    ucodeHeader->DateDay,
+                    ucodeHeader->DateMonth,
+                    ucodeHeader->DateYear);
+    realSize = ucodeHeader->TotalSize;
+    
+    return U_SUCCESS;
+}
+
+USTATUS FitParser::parseFitEntryAcm(const UByteArray & acm, const UINT32 localOffset, const UModelIndex & parent, UString & info, UINT32 &realSize)
+{
+    // Parse Intel ACM using Kaitai-generated parser
+    // TODO: need to do something with the copy here, probably implement a custom wrapper
+    std::istringstream is(std::string(acm.constData(), acm.size()));
```

Comment:
Ugh, will std::string work with \0 bytes? You can use something like this: https://tuttlem.github.io/2014/08/18/getting-istream-to-work-off-a-byte-array.html

Reply:
This is from Kaitai docs, and it explicitly does work with \0 bytes.
<img width="974" alt="Screenshot 2022-09-10 at 11 39 45" src="https://user-images.githubusercontent.com/1845688/189477830-e1204a75-18ce-4b14-a70e-caa7c14fcef7.png">